In [1]:
! apt-get update
! apt-get install mysql-server

Get:1 https://cli.github.com/packages stable InRelease [3,917 B]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:4 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:6 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages [85.2 kB]
Get:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:8 https://cli.github.com/packages stable/main amd64 Packages [357 B]
Get:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease [18.1 kB]
Get:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease [24.6 kB]
Get:11 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,785 kB]
Get:12 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:13 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [6,626 kB]
Ge

In [4]:
!mysql --version   # Making sure everything has been installed correctly by checking the version
!service mysql start   # Starting our server

mysql  Ver 8.0.45-0ubuntu0.22.04.1 for Linux on x86_64 ((Ubuntu))
 * Starting MySQL database server mysqld
   ...done.


In [3]:
# if the above block gave "su warning", run this block, then re-run the above block again
!sudo service mysql stop
!sudo usermod -d /var/lib/mysql/ mysql
!sudo service mysql start

 * Stopping MySQL database server mysqld
   ...done.
 * Starting MySQL database server mysqld
   ...done.


In [5]:
!wget https://www.kaggle.com/api/v1/datasets/download/davidgauthier/glassdoor-job-reviews

--2026-02-26 13:21:49--  https://www.kaggle.com/api/v1/datasets/download/davidgauthier/glassdoor-job-reviews
Resolving www.kaggle.com (www.kaggle.com)... 35.244.233.98
Connecting to www.kaggle.com (www.kaggle.com)|35.244.233.98|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://storage.googleapis.com:443/kaggle-data-sets/2545470/9631506/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20260226%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20260226T132149Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=4e64d16b194b33dfe73a58c76bf83d4d869f25abad4fb58e96b9876da60156715fe9e99cf747312563e0a0137d536aff7d6ef3b6b357ecbf6a745c8868b0de46f3a961da28806aca7d071bd8831f624c9562112407741aafee6118846530c59f1472c81d90bf9f86c53a74c7215769724c4e5a5f50cc41a4f20fa72d2f825a70412154575581fe5935b2baac956777f8c44bd991cabd3213569cae276f793605b5693d2068c11b651a68961eeaba0b13f14ada982

In [6]:
!unzip glassdoor-job-reviews

Archive:  glassdoor-job-reviews
  inflating: glassdoor_reviews.csv   


In [7]:
!wget https://www.kaggle.com/api/v1/datasets/download/davidgauthier/glassdoor-job-reviews-2

--2026-02-26 13:21:55--  https://www.kaggle.com/api/v1/datasets/download/davidgauthier/glassdoor-job-reviews-2
Resolving www.kaggle.com (www.kaggle.com)... 35.244.233.98
Connecting to www.kaggle.com (www.kaggle.com)|35.244.233.98|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://storage.googleapis.com:443/kaggle-data-sets/4677933/9631516/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20260226%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20260226T132155Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=301181c9a18f3fa24ffb6411dda3744ffe868d3d8811ee9e58160ba0ab4c80a53b85c0adceae2119dca486d635cc047e18ece5e4b0f2f9352e5f80e15f42ab1b9c575263b360deaa1f22d19fe08925efe03d5a6e2fe120ea8ba42d3a32c48d611b6036088d7940cf60050bca4884b1f958e9bb51de007530a7297e6564b4ed78a7a09a9d9612405a00c441a7ad19a995b7493e31b123bf1641ab841598bbc5c7896c14d3925a0d7a68e4a9e033e049131e5c8c4

In [8]:
!unzip glassdoor-job-reviews-2

Archive:  glassdoor-job-reviews-2
  inflating: all_reviews.csv         


In [42]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Data Cleaning

Below are a number of scripts to clean the datasets so that they can easily be inserted into the SQL database. First, we have to clone our GitHub repository which contains the python scripts we'll need

In [ ]:
!git clone https://github.com/DavidRemenyik/glassdoor-project.git

Cloning into 'glassdoor-project'...
fatal: could not read Username for 'https://github.com': No such device or address


Alternatively, if you have issues cloning the repository directly, you can save the python scripts to your mounted Google Drive and copy them to the local `content/` directory as below, replacing with your own path:

In [11]:
!cp -r /content/drive/MyDrive/"PATH_TO_PYTHON_FOLDER"/python /content/

## Dataset 1: `glassdoor_reviews.csv`

In [12]:
!time python3 python/clean-commas.py glassdoor_reviews.csv # remove commas, tabs, etc.


real	0m26.240s
user	0m25.158s
sys	0m0.601s


In [13]:
!cp output.csv dataset1.csv

In [14]:
!time tail -n +2 output.csv > dataset1-final.csv # remove the first row (headers)


real	0m1.091s
user	0m0.020s
sys	0m0.395s


## Dataset2: `all_reviews.csv`

In [15]:
!time python3 python/company_name_extractor.py

/content/python/company_name_extractor.py:3: DtypeWarning: Columns (5,9,10,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("all_reviews.csv")

real	6m24.831s
user	5m46.145s
sys	0m27.123s


In [16]:
!time python3 python/rename_columns.py name_extraction.csv


real	6m49.124s
user	6m15.599s
sys	0m17.834s


In [17]:
!time python3 python/reorder_columns.py renamed.csv


real	5m15.476s
user	4m49.627s
sys	0m15.120s


In [18]:
!time python3 python/clean-commas.py renamed_reordered.csv


real	6m11.863s
user	5m40.724s
sys	0m17.135s


In [19]:
!cp output.csv dataset2.csv

In [20]:
!time tail -n +2 output.csv > dataset2-final.csv # remove the first row (headers)


real	0m19.177s
user	0m0.306s
sys	0m5.827s


# Calculating Category Weights

With the cleaned datasets, we can now calculate the "weights" for each category we will use to determine our own overall rating. We calculate this via the following simple algorithm:

```python
for category, pattern in categories.items():
  pros = chunk["pros"].str.contains(pattern, regex=True).sum()
  cons = chunk["cons"].str.contains(pattern, regex=True).sum()
  
  weights[category] += pros - cons
```

Essentially, whenever we come across a word related to a given category in the "pros" column, we add +1 to that category's score. Similarly, if we find a word associated with a given category in the "cons" category, we subtract -1 from that category's score.

In [ ]:
!time python3 python/calculate-weights.py dataset1.csv

work_life_balance: 28171
culture_values: 95879
diversity_inclusion: 2788
career_opp: 10767
comp_benefits: 50552
senior_mgmt: -65912

real	0m46.457s
user	0m43.735s
sys	0m0.888s


In [ ]:
!time python3 python/calculate-weights.py dataset2.csv

work_life_balance: 287758
culture_values: 1025792
diversity_inclusion: 11032
career_opp: -45003
comp_benefits: 631741
senior_mgmt: -817900

real	9m12.885s
user	8m51.114s
sys	0m7.783s


From running the script on both datasets, we get the following total weights which we can use to weight our own rating system when we execute our SQL queries:


| Category           | dataset1 | dataset2 | Total |
| --------           | --- | --- | --- |
| work_life_balance  | 28171 | 287758 | 315929 |
| culture_values     | 95879 | 1025792 | 1121671 |
| diversity_inclusion| 2788 | 11032 | 13820 |
| career_opp         | 10767 | -45003 | -34236 |
| comp_benefits      | 50552 | 631741 | 682293 |
| senior_mgmt        | -65912 | -817900 | -883812 |

To convert these values to more appropriate floating point values:

$$
\text{weight_category} = total / \sum |total| \\ \\
|total| = 315929 + 1121671 + 13820 + 34236 + 682293 + 883812 = 2,639,761 \\
$$

$$
w1 =  315,929  / 2,639,761 =  0.1197 \\
w2 = 1,121,671 / 2,639,761 =  0.4247 \\
w3 =   13,820  / 2,639,761 =  0.0052 \\
w4 =  -34,236  / 2,639,761 = -0.0129 \\
w5 =  682,293  / 2,639,761 =  0.2584 \\
w6 = -883,812  / 2,639,761 = -0.3348 \\
$$


# SQL Queries

Now that we have our cleaned datasets, we can insert them into our SQL database and start running queries on them to extract information from them

In [21]:
!mysql -u root -e "SET GLOBAL local_infile = 1;" # Need to allow file loading, if prompted for a password just press ENTER (leave blank)

In [22]:
!mysql -e "CREATE DATABASE glassdoorReviews;"

In [23]:
!mysql -e "USE glassdoorReviews; CREATE TABLE employee (firm TEXT, date_review VARCHAR(20), job_title TEXT, current TEXT, location TEXT, overall_rating FLOAT, work_life_balance FLOAT, culture_values FLOAT, career_opp FLOAT, comp_benefits FLOAT, senior_mgmt FLOAT, recommend VARCHAR(20), ceo_approv VARCHAR(20), outlook VARCHAR(20), headline TEXT, pros TEXT, cons TEXT);"

In [24]:
!time mysql --local-infile=1 -D glassdoorReviews -e " \
LOAD DATA LOCAL INFILE '/content/dataset1-final.csv' \
INTO TABLE employee \
FIELDS TERMINATED BY ',' \
OPTIONALLY ENCLOSED BY '\"' \
LINES TERMINATED BY '\n' \
IGNORE 1 ROWS \
(firm, date_review, job_title, current, location, \
 overall_rating, work_life_balance, culture_values, \
 career_opp, comp_benefits, senior_mgmt, recommend, \
 ceo_approv, outlook, headline, pros, cons); \
"



real	0m29.544s
user	0m0.057s
sys	0m0.377s


In [25]:
!time mysql --local-infile=1 -D glassdoorReviews -e " \
LOAD DATA LOCAL INFILE '/content/dataset2-final.csv' \
INTO TABLE employee \
FIELDS TERMINATED BY ',' \
OPTIONALLY ENCLOSED BY '\"' \
LINES TERMINATED BY '\n' \
IGNORE 1 ROWS \
(firm, date_review, job_title, current, location, \
 overall_rating, work_life_balance, culture_values, \
 career_opp, comp_benefits, senior_mgmt, recommend, \
 ceo_approv, outlook, headline, pros, cons); \
"


real	6m31.324s
user	0m0.516s
sys	0m4.269s


In [26]:
!time mysql -e "USE glassdoorReviews; SELECT COUNT(*) FROM employee;"

+----------+
| COUNT(*) |
+----------+
| 10740378 |
+----------+

real	0m17.604s
user	0m0.008s
sys	0m0.005s


In [27]:
!time mysql -e "USE glassdoorReviews; SELECT * FROM employee LIMIT 1\G;"

*************************** 1. row ***************************
             firm: AFH-Wealth-Management
      date_review: 2015-12-11
        job_title: Office Administrator
          current: Current Employee more than 1 year
         location: Bromsgrove England England
   overall_rating: 2
work_life_balance: 3
   culture_values: 1
       career_opp: NULL
    comp_benefits: 2
      senior_mgmt: 1
        recommend: 4.0
       ceo_approv: x
          outlook: o
         headline: r
             pros: Excellent staff poor salary
             cons: Friendly helpful and hard-working colleagues

real	0m0.062s
user	0m0.004s
sys	0m0.008s


In [28]:
!time mysql -e "USE glassdoorReviews; SELECT FIRM, COUNT(*) FROM employee GROUP BY FIRM ORDER BY COUNT(*) DESC LIMIT 10;"

+--------------------------------+----------+
| FIRM                           | COUNT(*) |
+--------------------------------+----------+
| Amazon                         |   163396 |
| McDonald-s                     |   126227 |
| IBM                            |   112508 |
| Tata-Consultancy-Services      |   107218 |
| Walmart                        |   102152 |
| Deloitte                       |    90757 |
| PwC                            |    84413 |
| Cognizant-Technology-Solutions |    84171 |
| Oracle                         |    75146 |
| Accenture                      |    71182 |
+--------------------------------+----------+

real	1m46.710s
user	0m0.006s
sys	0m0.005s


In [29]:
# The firms with the highest rating (with at least 1000 reviews)
!time mysql -e " \
USE glassdoorReviews; SELECT \
    firm, \
    AVG(overall_rating) AS avg_rating, \
    COUNT(*) AS review_count \
FROM employee \
WHERE overall_rating IS NOT NULL \
GROUP BY firm \
HAVING COUNT(*) >= 1000 \
ORDER BY avg_rating DESC \
LIMIT 10; \
"

+--------------------------------+--------------------+--------------+
| firm                           | avg_rating         | review_count |
+--------------------------------+--------------------+--------------+
| Universitas-Gadjah-Mada        |  4.589792060491494 |         4232 |
| Bain-and-Company               |  4.544142143253748 |         9005 |
| Continental-Software-Solutions |  4.499505440158259 |        24264 |
| HubSpot                        |  4.488389513108614 |         2670 |
| Brigham-Young-University-BYU   |  4.485022466300549 |         4006 |
| Brigham-Young-University-(BYU) |    4.4835297673013 |        19854 |
| Carnegie-Mellon-University     |  4.439096267190569 |         2036 |
| Zoho                           | 4.4164296998420225 |         3165 |
| Google                         |  4.407604299087022 |        43265 |
| AIESEC                         |  4.400535833891493 |         1493 |
+--------------------------------+--------------------+--------------+

real	

In [30]:
# The firms with the most reviews and their average ratings
!time mysql -e " \
USE glassdoorReviews; SELECT \
    firm, \
    COUNT(*) AS review_count, \
    AVG(overall_rating) AS avg_rating \
FROM employee \
WHERE overall_rating IS NOT NULL \
GROUP BY firm \
ORDER BY review_count DESC \
LIMIT 10; \
"

+--------------------------------+--------------+--------------------+
| firm                           | review_count | avg_rating         |
+--------------------------------+--------------+--------------------+
| Amazon                         |       163393 | 3.7307106179579295 |
| McDonald-s                     |       126226 |  3.329892415191799 |
| IBM                            |       112508 | 3.6044014647847264 |
| Tata-Consultancy-Services      |       107216 | 3.7840247724220264 |
| Walmart                        |       102151 |  3.189151354367554 |
| Deloitte                       |        90757 | 3.8093149839681786 |
| PwC                            |        84411 |  3.787918636196704 |
| Cognizant-Technology-Solutions |        84169 | 3.7117941284796063 |
| Oracle                         |        75145 | 3.5146050968128284 |
| Accenture                      |        71179 |  3.847230222397055 |
+--------------------------------+--------------+--------------------+

real	

In [31]:
# The firms with the most reviews and their average ratings (with our weights)
!time mysql -e " \
USE glassdoorReviews; WITH top10 AS ( \
    SELECT firm \
    FROM employee \
    GROUP BY firm \
    ORDER BY COUNT(*) DESC \
    LIMIT 10 \
) \
SELECT \
    e.firm, \
    COUNT(*) AS review_count, \
    (AVG(\
          e.work_life_balance   *  1.1197 \
        + e.culture_values      *  1.4247 \
        + (e.career_opp          + (e.career_opp * -0.0129)) \
        + e.comp_benefits       *  1.2584 \
        + (e.senior_mgmt         + (e.senior_mgmt * -0.3348)) \
    ) / 5) AS weighted_overall_rating \
FROM employee e \
JOIN top10 t ON e.firm = t.firm \
GROUP BY e.firm \
ORDER BY review_count DESC; \
"

+--------------------------------+--------------+-------------------------+
| firm                           | review_count | weighted_overall_rating |
+--------------------------------+--------------+-------------------------+
| Amazon                         |       163396 |       4.035940079162847 |
| McDonald-s                     |       126227 |      3.5615688849088314 |
| IBM                            |       112508 |       4.344038401993682 |
| Tata-Consultancy-Services      |       107218 |       4.138706556905689 |
| Walmart                        |       102152 |      3.4748539708891393 |
| Deloitte                       |        90757 |       4.084184941566868 |
| PwC                            |        84413 |       4.046851913106837 |
| Cognizant-Technology-Solutions |        84171 |      3.8922512711387673 |
| Oracle                         |        75146 |       4.040865176254544 |
| Accenture                      |        71182 |       4.218110203381628 |
+-----------

In [32]:
# The firms with the highest rating (with at least 1000 reviews, with our weights)
!time mysql -e " \
USE glassdoorReviews; \
SELECT \
    firm, \
    COUNT(*) AS review_count, \
    (AVG(\
          work_life_balance   *  1.1197 \
        + culture_values      *  1.4247 \
        + (career_opp          + (career_opp * -0.0129)) \
        + comp_benefits       *  1.2584 \
        + (senior_mgmt         + (senior_mgmt * -0.3348)) \
    ) / 5) AS weighted_rating \
FROM employee \
WHERE work_life_balance IS NOT NULL \
  AND culture_values IS NOT NULL \
  AND career_opp IS NOT NULL \
  AND comp_benefits IS NOT NULL \
  AND senior_mgmt IS NOT NULL \
GROUP BY firm \
HAVING COUNT(*) >= 1000 \
ORDER BY weighted_rating DESC \
LIMIT 10; \
"


+--------------------------------+--------------+--------------------+
| firm                           | review_count | weighted_rating    |
+--------------------------------+--------------+--------------------+
| Continental-Software-Solutions |         4852 |  5.356390000000444 |
| Universitas-Gadjah-Mada        |         2526 |   4.84269431512287 |
| Globant                        |         1650 |  4.788490981818208 |
| HubSpot                        |         1180 |  4.786479711864408 |
| ServiceNow                     |         1850 |  4.772648118918962 |
| Zoho                           |         1459 |  4.755981905414659 |
| Adobe                          |         1604 | 4.7349211471321855 |
| Google                         |        12028 |  4.724546288659423 |
| Vector-Marketing               |         1666 |  4.716408247298945 |
| TaskUs                         |         3286 |  4.710303907486434 |
+--------------------------------+--------------+--------------------+

real	

# Spark

This section focuses on our use of Spark to speed up our query processes.

In [43]:
!cp -r /content/drive/MyDrive/Programming-for-Big-Data-Project-Files/spark/ /content/

In [1]:
%cd spark/

/home/david/Documents/Year3Spring/BigData/glassdoor-project/spark


In [45]:
# !python3 pySparkBasic.py

python3: can't open file '/content/spark/pySparkBasic.py': [Errno 2] No such file or directory


In [ ]:
!cat ../dataset1-final.csv ../dataset2-final.csv > ../mergedFinal.csv

In [2]:
!time python3 calculate-weights-spark.py ../dataset1-final.csv

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
26/02/26 14:48:03 WARN Utils: Your hostname, david-HP-ProBook-x360-440-G1, resolves to a loopback address: 127.0.1.1; using 172.18.230.109 instead (on interface wlp2s0)
26/02/26 14:48:03 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/02/26 14:48:04 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
work_life_balance: -47119                                                                                                                                
culture_values: -119272                                                                                                                                  


In [4]:
!time python3 calculate-weights-spark.py ../dataset2-final.csv

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
26/02/26 14:51:17 WARN Utils: Your hostname, david-HP-ProBook-x360-440-G1, resolves to a loopback address: 127.0.1.1; using 172.18.230.109 instead (on interface wlp2s0)
26/02/26 14:51:17 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/02/26 14:51:18 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
work_life_balance: -426692                                                                                                                               
culture_values: -1280280                                                                                                                                 


In [5]:
!time python3 sql-query-spark.py ../dataset2-final.csv

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
26/02/26 15:01:14 WARN Utils: Your hostname, david-HP-ProBook-x360-440-G1, resolves to a loopback address: 127.0.1.1; using 172.18.230.109 instead (on interface wlp2s0)
26/02/26 15:01:14 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/02/26 15:01:15 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable

--- TOP 10 WEIGHTED COMPANIES ---
+------------------------------+------------+------------------+                                                                                         
|firm                          |review_count|weighted_rating   |
+------------------------------+------------+---------

In [3]:
!time python3 sql-query-spark.py ../mergedFinal.csv

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
26/02/26 14:49:21 WARN Utils: Your hostname, david-HP-ProBook-x360-440-G1, resolves to a loopback address: 127.0.1.1; using 172.18.230.109 instead (on interface wlp2s0)
26/02/26 14:49:21 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/02/26 14:49:23 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable

--- TOP 10 WEIGHTED COMPANIES ---
+------------------------------+------------+------------------+                                                                                         
|firm                          |review_count|weighted_rating   |
+------------------------------+------------+---------